### ISSUE TITLE
Xybots (and clones) missing RetroPad mapping

### ISSUE THREAD

**Issue by paradadf** [libretro/mame2003-libretro#69]
Xybots should have action buttons (twist) mapped to RetroPad buttons by default and not only key strokes.

![20161214_170937544_ios](https://cloud.githubusercontent.com/assets/15956550/21193372/5b6ce5a8-c22c-11e6-8164-6351e84d1772.jpg)

**Comment by dankcushions** [comment]
good spot! 

https://github.com/libretro/mame2003-libretro/blob/7b55725e744ac9702db092f996e2fca2276146b2/src/drivers/xybots.c#L122

it seems that this and many other games with inputs that use `IP_JOY_DEFAULT` are not actually bound to anything on the joypad. it's almost like a placeholder... very strange.

for this game, what would be good retropad buttons to use for this game? the original cabinet had a 'twist' stick. i think retropad Y and A as this maps well to the right_stick_as_btns core option, allowing you to the use the right stick to rotate the view, sort of like a modern FPS/3PS.

**Comment by paradadf** [comment]
Yes, A and Y would be perfect. That's the most intuitive I would say.

**Comment by markwkidd** [comment]
Is this working now?

**Comment by grant2258** [comment]
T_BITX(0x0004, IP_ACTIVE_LOW, IPT_BUTTON3 | IPF_PLAYER2, "P2 Twist Right", KEYCODE_W, IP_JOY_DEFAULT ) 

this defaults to button 3

**Comment by grant2258** [comment]
Thats one thing i never did understand why not just bind the buttons number to the retropad

**Comment by dankcushions** [comment]
it is? https://github.com/libretro/mame2003-libretro/blob/master/src/libretro/joystick.c#L20

i think the issue is with this IP_JOY_DEFAULT thing - it's avoiding this association for whatever reason. i never investigated beyond that.

**Comment by grant2258** [comment]
{ IPT_BUTTON3             | IPF_PLAYER1, "P1 Button 3",    SEQ_DEF_5(KEYCODE_SPACE, CODE_OR, JOYCODE_1_BUTTON3, CODE_OR, JOYCODE_MOUSE_1_BUTTON2) },

**Comment by grant2258** [comment]
code snippet

note ( IP_JOY_DEFAULT == CODE_DEFAULT)

	if (seq_get_1((InputSeq*)&in->seq) != CODE_DEFAULT)
		return (InputSeq*)&in->seq;

	i = 0;

	while (inputport_defaults[i].type != IPT_END &&
			inputport_defaults[i].type != type)
		i++;

	if ((in->type & ~IPF_MASK) == IPT_EXTENSION)
		return &inputport_defaults[i+1].seq;
	else
		return &inputport_defaults[i].seq;
}

**Comment by grant2258** [comment]
the code reads this way to me ill need to verify though

**Comment by dankcushions** [comment]
i'm sure that's how it works in standalone, but i'm saying it's not working in the libretro version for some reason. retropad's button3 mapping is being ignored when in IP_JOY_DEFAULT  situations

**Comment by grant2258** [comment]
@dankcushions maybe we are setting the binding too soon might be worth fixing at core level rather than every driver

**Comment by dankcushions** [comment]
yeah my hunch is that this is an order thing. i don't like the idea of changing the inputs in the drivers for any reason.

**Comment by grant2258** [comment]
@dankcushions  xybots doesnt map to a joypad in mame078

however look here in 
alpha68k.c

dame paddlema

	PORT_BITX(0x0100, IP_ACTIVE_LOW, IPT_START1, "Button A (Start)", IP_KEY_DEFAULT, IP_JOY_DEFAULT )
	PORT_BITX(0x0200, IP_ACTIVE_LOW, IPT_START2, "Button B (Start)", IP_KEY_DEFAULT, IP_JOY_DEFAULT )

this code works maybe this setting is reserved for service,start ect ill need to look up the code more it might just be fixable with gfag changes to use the right retropad buttons

**Comment by SapphireTactician** [comment]
@markwkidd I'll download the new code and recompile the core, then take a look into it for you.

**Comment by SapphireTactician** [comment]
@markwkidd Downloaded latest code, compiled, twist controls still not mapped to RetroPad.

![2018-04-24--135641_1600x900](https://user-images.githubusercontent.com/24427487/39205357-64fcc9e4-47c8-11e8-8dcd-07e26d3cea94.png)

**Comment by SapphireTactician** [comment]
@markwkidd I figured out what happened. It's not pretty. To get the new defaults you assigned for each twist had to "null" out those controls first. Meaning each twist can't be truly nulled anymore as nulling sets them to what you intended to be the defaults.

**Comment by markwkidd** [comment]
@paradadf and @SapphireTactician -- @grant2258 added remappable analog inputs to MAME 2003-Plus yesterday.

That means that analog inputs can be remapped from within the frontend (RetroArch) rather than from within the MAME Menu/Remapper.

If either of you is still interested in this game, would you be willing to give it a try with the latest MAME 2003-Plus build, and try to map the twister completely from within the frontend?

**Comment by paradadf** [comment]
@markwkidd I just tried to remapp the missing inputs from within retroarch‘s controls menu and had no luck. I couldn‘t map anything, no response.

**Comment by grant2258** [comment]
The inputs where only added to the mame mapping and only on plus. youll need to use the tab menu unless someone adds code to add these finger for ra right now it just generic returns and wrong in some cases(the ra end).

**Comment by paradadf** [comment]
I used the plus core as markwkidd asked for.

**Comment by grant2258** [comment]
Im not sure where mark is getting his wires crossed its just confusing people. 

Lost of games have issues with  the frontend mapping anything analog or not mapped to a button by default. like zwakery, birdtry, spacewars ect you need to use the tab menu. 

Everything is settable in the tab menu all you can do in the frontend is remap whats mapped to default not all drivers map to default though and marks front end impemention only supports digital atm.

**Comment by rsn8887** [comment]
How can I remove a Mame input menu mapping on a system without a keyboard? I think on PC it was „esc“.

**Comment by grant2258** [comment]
enable legacy mode and press l3

**Comment by grant2258** [comment]
https://github.com/libretro/mame2003-plus-libretro/blob/c6fc83a0db96e5065db8be26b194fb309357c84e/src/inptport.c#L407

**Comment by grant2258** [comment]
https://github.com/libretro/mame2003-libretro/blob/133ef799e03b29001ca274f17c5ab43558e28dd2/src/inptport.c#L405

**Comment by rsn8887** [comment]
Interesting. I wish Mame osd menu had a little reminder text how to do this, since it is impossible to figure it out without instructions.

**Comment by grant2258** [comment]
https://docs.libretro.com/library/mame_2003/

**Comment by rsn8887** [comment]
Nice docs! But the "press left twice to return to default" is not mentioned. That's the part I find impossible to figure out.
